Get, save, and load module information for each operator

taivu1998 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42133 Get, save, and load module information for each operator**
* #41615 Reconstruct scopes

Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

Commit:
Get, save, and load module information for each operator

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"




[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"




[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"




[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"




[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"




[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"




[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"




[ghstack-poisoned]

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -1,83 +1,127 @@
   }
 }
 
+std::string ReconstructScopesPass::getSubModuleName(
```
Just call it `getModuleName`?

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -1,83 +1,127 @@
+  module_names[module._ivalue()] = getSubModuleName(module, prefix);
   for (NameModule s : module.named_children()) {
-    constructRelativeNamesForModules(s.value, prefix + "." + s.name);
+    std::string newPrefix = (!has_duplicated_function) ?
```
Instead of calling has_duplicated_function, maybe we can say class_types_are_not_unique. Also put a comment suggesting why is this necessary? Give an example, like when doing traced module vs scripted module with two instances of same type what can happen, and what we are doing for that.

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -1,83 +1,127 @@
   for (NameModule s : module.named_children()) {
-    constructRelativeNamesForModules(s.value, prefix + "." + s.name);
+    std::string newPrefix = (!has_duplicated_function) ?
+        module_names[module._ivalue()] + "." + s.name : module_names[module._ivalue()] + ".";
```
When has_duplicated_function=true, I suggest add something like <instance-name-unknown>.

taivu1998 replied:
Can you clarify your suggestion a bit? Also, since we decided not to save module info the two cases differently, I believe that we don't need to check for `has_duplicated_function` here. Is that correct?

kimishpatel replied:
> Can you clarify your suggestion a bit? Also, since we decided not to save module info the two cases differently, I believe that we don't need to check for `has_duplicated_function` here. Is that correct?

Are you talking about not checking `has_duplicated_function` for this PR or after the other PR

kimishpatel replied:
> When has_duplicated_function=true, I suggest add something like .

My first comment had trail off. I meant when has_duplicated_function=true, add ".<instance info unavailable>" or something like that.

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -1,83 +1,127 @@
-  return module_names.at(m) + "." + f->name();
+  std::string scopeString = module_names.at(m) + "." + f->name();
+
+  if (has_duplicated_function) {
+    SourceRange r = frame.second;
+    if (r.source()) {
+      if (auto orig = r.source()->findSourceRangeThatGenerated(r)) {
+        r = *orig;
+      }
+    }
+    if (auto file_line_col = r.file_line_col()) {
+      std::string filename;
+      size_t line, col;
+      std::tie(filename, line, col) = *file_line_col;
+      scopeString += "(" + filename + ":" + c10::to_string(line) + ":" + c10::to_string(col) + ")";
+    }
+  }
```
Not clear to me why we should do this? If we decide that this might help, then it is probably best to split it out into a helper function with comments and fn name that suggests what it will do.
But let's first clarify what does this do.

taivu1998 replied:
That was for adding source range info in the case of a scripted module, as @ZolotukhinM suggested before. But I think I'm gonna remove it now as we decided not to do things differently for the two types of modules.

kimishpatel replied:
Make it a separate function in that case, atleast, till we resolve it.

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -89,7 +133,13 @@ void ReconstructScopesPass::run() {
 
-  visitBlock(graph->block());
+  if (has_duplicated_function) {
+    std::cout << "WARNING: There are some duplicated module instances." << std::endl;
```
This is not clear. Maybe say your module seems to have two instances of the same type. Current debugging does not support distinguishing the two instances of the same type yet. Thus debugging information maybe incomplete. Or something to the effect.

taivu1998 replied:
Yes, I'll fix that. One small question: Since it's a warning, is it ok to directly print it directly to std::cout like this?

kimishpatel replied:
Good question. I missed this. I think there is something like `TORCH_WARN` or torch_debug soemthing. I will have to check.

kimishpatel commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -134,38 +159,55 @@ c10::IValue getFunctionTuple(const Function& func) {
+      module_paths.emplace_back(std::move(path));
+    }
+
+    auto debug_info = Table({{"module_debug_info", Tup(module_paths)}});
```
Call this op_debug_info. The debug info contains module information but, debug info itself is for ops.

kimishpatel commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -134,38 +159,55 @@ c10::IValue getFunctionTuple(const Function& func) {
                       {"types", Tup(types)},
```
It should be saved as part of this table, not separately.

taivu1998 replied:
Can you clarify it a little bit, because I don't think I made any changes to this line?

kimishpatel replied:
I mean your change saving it outside the main table. As as a new element in tuple. I am saying we should combine this with main table. I can help clarify if this is not clear.

kimishpatel commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -134,38 +159,55 @@ c10::IValue getFunctionTuple(const Function& func) {
 
   // __setstate__ of all components
-  setstateTuple(module._ivalue(), elements);
+  setstateTuple(module, module._ivalue(), elements, save_debug_info);
```
I am wondering if we should do `    ReconstructScopes(module, *graph, "top");` here and avoid passing module around. I think it makes sense to do it here. It is effectively saying, let's populate some debug info. And I think it does not need to be done closest to where it will be saved.

taivu1998 replied:
I thought about it yesterday. I'm not sure whether we can call `ReconstructScopes` here, because the `graph` did change each time we call the function. However, I was thinking that we can perhaps move `constructFunctionToModuleMap` to this file and only pass that map to `ReconstructScopes` (because our `module` stay unchanged, so it doesn't make sense to construct that map again and again in `ReconstructScopes`. What do you think?

kimishpatel replied:
Aah nevermind. Yeah we need to do this after inlining and not before. Let's see.

kimishpatel commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -357,37 +400,40 @@ void ExportModule(
 void export_opnames(const script::Module& m, std::set<std::string>& opnames) {
   std::vector<c10::IValue> elements;
-  moduleMethodsTuple(m, elements);
+  moduleMethodsTuple(m, elements, false /* save_debug_info */);
```
Good putting a comment there.

taivu1998 replied:
Just a quick question. Here I assumed that we didn't need to save debug info when calling `export_opnames`. Was that correct?

kimishpatel replied:
I think so. @iseeyuan?

kimishpatel commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -357,37 +400,40 @@ void ExportModule(
     const ExtraFilesMap& extra_files,
-    bool bytecode_format) {
+    bool bytecode_format,
+    bool save_debug_info) {
```
Actually given that we are saving this only for byte code right now, lets call it save_debug_info_in_bytecode.

kimishpatel commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -125,7 +130,20 @@ void parseMethods(
     const auto& types_list =
         expect_field(table, "types", BYTECODE_INDEX_TYPE).toTuple()->elements();
-    const auto& register_size = expect_field(table, "register_size", 4).toInt();
+    const auto& register_size = expect_field(table, "register_size", BYTECODE_INDEX_REGISTER_SIZE).toInt();
```
Thanks for this change.

kimishpatel commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -140,16 +158,19 @@ void parseMethods(
       if (!op_found) {
         unsupported_op_names.emplace(operator_str(
-            op_item[0].toString()->string(), op_item[1].toString()->string()));
+            op_item[0].toString()->string(), op_item[1].toString()->string(), module_debug_info));
```
Since this function is used only for unsupported op names, let's not add module info to that. The error message out of this becomes too long and unclear. unsupported_op_names is mainly used to indicate what ops are not supported. No need to add debug info there.

taivu1998 replied:
That's good to know. I'll fix that.

kimishpatel commented on torch/csrc/jit/api/module.h:
```diff
@@ -213,19 +213,23 @@ struct TORCH_API Module : public Object {
       const std::string& filename,
-      const ExtraFilesMap& extra_files = ExtraFilesMap()) const;
+      const ExtraFilesMap& extra_files = ExtraFilesMap(),
+      bool save_debug_info = false) const;
```
We should not add this to anything other than `_save_for_mobile` as that is what we are doing. Please revert these ones.

taivu1998 replied:
Thanks for catching it! I'll fix that.

kimishpatel commented on torch/csrc/jit/mobile/interpreter.cpp:
```diff
@@ -52,7 +55,10 @@ bool InterpreterState::run(Stack& stack) {
           enableRecordFunction(true);
         }
-        RECORD_FUNCTION(code_->op_names_[inst.X].name, stack);
+        std::string fn_name = (!module_debug_info_list_[inst.X].empty()) ?
```
I think we should not do this here. String manipulation is critical path will add latency. Given that op_index corresponds to pc, we can add a function in `Function` that constructs a map/vector <pc, debug_string>, probably vector is better and have an API provided at mobile::Module level where you can query debug info for a PC. This can then be used in predictor to augment debug info to op names. This way getting debug info can be optional and if a client does not want debug info, it does not get it. @iseeyuan, what do you think?

taivu1998 replied:
For this, is it ok to save op names augmented with module info (like `top.A1.aten::add`) in `Function` instead of saving module info (like `top.A1`). This way, we don't have to do string concatenation when running our model. Or should we save both of them in two separate vectors?

kimishpatel replied:
That is what we wanted to avoid because now interpreter has to reverse engineer to find op name if you append it like that. I think the best thing to do is: After having done append_instruction and append_operator, have a function that iterates over instructions, where each index is PC, find inst[PC].X which is op_index, find corresponding module_debug_info and then instead of having module_debug_info_list_ that has one-to-one correspondence with operator list, make it map<PC, module_debug_info>. Now you can add a function `get_module_debug_info(size_t PC)` which let's you query a Function. This can be exposed via mobile::Module by adding a method `get_forward_method_debug_info(size_t PC)`. So then in predictor when we finally print out op with latency info, we can query debug info using PC which we had also used for mapping op_names.

kimishpatel reviewed (changes requested):
Left some comments. Address these and add unit tests as we discussed.

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/mobile/module.cpp:
```diff
@@ -129,6 +129,11 @@ const std::map<std::string, at::Tensor> Module::named_parameters() const {
   return params;
 }
+
+std::string Module::get_forward_method_debug_info(size_t pc) {
```
This method seems to be only used in tests and doesn't look general enough to be in the Module API. Would it make sense to convert it to a helper function in the test?

taivu1998 replied:
I think we do need it to display module info when performing latency profiling. @kimishpatel

kimishpatel replied:
@ZolotukhinM, so the idea right now is this: For each PC, there is a corresponding debug info available that the module allows you to query wrt its forward method. So then in lite predictor that uses loads `mobile::Module` and runs the forward method of it, can query the current PC and get corresponding debug info.
This is honestly not super clean, but the existing operator profiling infra was built this way. https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/mobile/interpreter.cpp#L43. Definitely not ideal, but I think we will have to fix that in longer term. I would prefer not having Tai fix issues we introduced ourselves :)

kimishpatel replied:
On the other hand, I think we can do something like this. We can ask lite interpreter to give debug info of current pc, so like `get_current_pc_debug_info()`, that returns you the string.
Then in your test code, as well as in lite predictor, as we run the model, we can have onFunctionEnter/Exit calls that get called as if when we are doing profiling. And within this call you can collect all the debug info and check against what you expect. And you do this only once.
```
in OnFunctionEnter
if (debug_info_wanted) {
debug_info = module.get_current_pc_debug_info()
append debug info to vector.
Only in first forward method call debug info is extracted and for later runs
get_debug_info = false.
}
```

ZolotukhinM commented on torch/csrc/jit/mobile/function.h:
```diff
@@ -2,6 +2,7 @@
 #include <ATen/core/ivalue.h>
 //#include <aten/src/Aten/core/operator_name.h>
 #include <vector>
+// #include <unordered_map>
```
Nit: a leftover.

ZolotukhinM commented on torch/csrc/jit/mobile/interpreter.h:
```diff
@@ -18,13 +18,16 @@ struct Code {
   std::shared_ptr<Code> code_;
   c10::IValue& reg(size_t reg);
   std::vector<c10::IValue> registers_;
+  std::vector<std::string> module_debug_info_list_;
```
I don't see how it is used in Interpreter. Could you please elaborate on that?

taivu1998 replied:
Oh It was from my previous version. Thanks for catching that. I'm gonna remove it.

ZolotukhinM commented on torch/csrc/jit/mobile/function.h:
```diff
@@ -21,14 +22,21 @@ class Function {
   c10::QualifiedName name_;
   std::shared_ptr<Code> code_;
+  std::vector<std::string> module_debug_info_list_;
+  // std::unordered_map<size_t, std::string> pc_to_module_debug_info_;
```
Nit: a leftover.

ZolotukhinM commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -45,21 +48,43 @@ static IValue Table(
   Inline(*graph);
-  torch::jit::Code code(graph, func.name());
+  if (save_debug_info_in_bytecode) {
+    ReconstructScopes(module, *graph, "top");
```
I would probably prefer if we serialize the source of truth, i.e. the InlinedCallStacks, rather than a derivative from it, which is the scopes we reconstruct here.

kimishpatel replied:
So, fundamentally I agree. This will also reduce the bloat in model size as we will be saving a tree rather than every visited path. My plan was to do this as a follow-up diff and let Tai continue his work as it may end up being broader in scope than what his internship allows.

kimishpatel replied:
@iseeyuan, @ZolotukhinM are you guys ok with this plan for now? This will definitely blow up model size, but we should do follow-up diffs to manage that better.

iseeyuan replied:
Looks fine for me for the debug info.

ZolotukhinM reviewed (commented):
I left some comments inline, but probably @suo is a better reviewer for the serialization changes.

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

iseeyuan reviewed (commented):
Could you add the debug info part to the example at the beginning of the mobile/import.cpp?

taivu1998 commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -53,6 +53,14 @@ bool Function::append_operator(
+}
+
+void Function::append_module_info(const std::string& module_info, size_t pc) {
+  pc_to_module_debug_info_[pc] = module_info;
```
Thanks for catching that!

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

kimishpatel commented on test/cpp/jit/tests.h:
```diff
@@ -8,97 +8,101 @@
 
 namespace torch {
 namespace jit {
-#define TH_FORALL_TESTS(_)             \
```
Did you add something here? Why is there this change?

kimishpatel commented on test/test_jit.py:
```diff
@@ -1222,7 +1222,7 @@ def forward(self, x):
         torch._C._jit_pass_inline(g)
         torch._C._jit_pass_reconstruct_scopes(traced._c, g)
-        FileCheck().check("scope: top.sub.forward").run(g)
+        FileCheck().check("scope: top(MyModule).sub(SubModule).forward").run(g)
```
Keep reconstruct scope related changes in base diff.

taivu1998 replied:
That's what I intended to do, but things like `top(MyModule).sub(SubModule).forward` can't be tested on the base diff. Sorry about that.

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

kimishpatel commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -65,10 +76,18 @@ void Function::set_register_size(size_t size) {
 }
 
+std::string Function::get_module_debug_info(size_t pc) {
+  if (pc >= pc_to_module_debug_info_.size()) {
```
Let's just assert here.

kimishpatel replied:
Or rather assert if we have module debug info available but pc is larger.
And return "no module info" if debug info is unavailable.

kimishpatel commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -53,6 +53,17 @@ bool Function::append_operator(
   return true;
 }
 
+void Function::set_module_debug_info_list_size(size_t size) {
+  pc_to_module_debug_info_.resize(size);
+}
```
I think it is ok to just do push_back. We dont expect to do this too often. We should aim to as less of a code as possible, unless needed. Less coding is better coding, I think.

taivu1998 replied:
Yes, but I don't think we can do `push_back` here because the vector may be sparse, i.e. we don't have module info for every `pc` value.

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -7,77 +7,163 @@ namespace jit {
   Graph& graph_;
   std::string prefix_;
 
+  // This boolean indicates whether there are two submodules of the same
```
two objects.

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -7,77 +7,163 @@ namespace jit {
+  // This boolean indicates whether there are two submodules of the same
+  // class type. This issue may occur in q scripted module and make it
+  // difficult to exactly track module information corresponding to each
+  // Node* after inlining the graph. Consider the following example:
```
I think we should have kept all the changes to reconstruct scopes in base diff. It makes it easy to revert a particular commit.

taivu1998 replied:
Makes sense. I think I'll move every change to `reconstruct_scopes.cpp` from the top diff to the base diff.

kimishpatel replied:
Thanks Tai. I understand this is a bit cumbersome, but it looks clean.

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -7,77 +7,163 @@ namespace jit {
   }
 }
 
+std::string ReconstructScopesPass::getModuleName(
```
Rename this to `getModuleTypeName`

kimishpatel commented on torch/csrc/jit/passes/reconstruct_scopes.cpp:
```diff
@@ -7,77 +7,163 @@ namespace jit {
-  return module_names.at(m) + "." + f->name();
+  std::string scopeString = module_names.at(m) + "." + f->name();
+
+  if (class_types_are_not_unique_) {
```
Put a comment here describing why we do this only if class types are not unique. Mainly to provide a little more information in the debug.

kimishpatel commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -44,21 +47,43 @@ static IValue Table(
+std::string getModulePath(Node* node) {
+  std::string modulePath = node->scopeName();
+  size_t end = modulePath.size();
+  if (modulePath[end - 1] == '>') {
```
I dont quite follow what is happening here. Are you trying to extract just the filename? Also do note that scopeName contains multiple scopes which are all appended together, right? In that case you are touching only the last file name, isnt it? Instead of doing this, why not just save the filename in the first place? Or just keep the whole path.

taivu1998 replied:
Here, I removed the source range info, which is displayed as `<filename:line:col>`, because we agreed to not display things differently for traced vs. scripted modules. But we can talk more today to figure it out.

taivu1998 replied:
Add a comment to explain this.

kimishpatel commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -137,6 +155,12 @@ void parseMethods(
+        std::string module_debug_info = (hasDebugInfo)
+            ? module_debug_info_list[X].toString()->string()
+            : "";
+        function->append_module_info(module_debug_info, i);
```
Maybe call this `set_module_info` since we are setting and not appending.

kimishpatel commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -300,6 +302,173 @@ void testLiteInterpreterBuiltinFunction() {
+  size_t pc = 0;
+  while (true) {
+    std::string module_info = bc.get_forward_method_debug_info(pc);
+    if (module_info == "<no module info for pc>") {
```
I think we can do testing as I described below where instead of using pc to extract this we can just ask do `bc.get_current_pc_debug_info()`. Check my comment later.

kimishpatel replied:
I think for now you can just to `try.. catch` clause and populate it that way. We will have to fix the interface more appropriately later.

kimishpatel commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -300,6 +302,173 @@ void testLiteInterpreterBuiltinFunction() {
+
+  std::unordered_set<std::string> expected_result(
+      {"top(C).A0(A).forward", "top(C).B0(B).forward"});
+  AT_ASSERT(module_debug_info_set == expected_result);
```
Add 
- Where we get top(C).A0(A).forward.B0(b).forward.
- And also one where you have A0 and A1 and then you can get only type information.

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

kimishpatel commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -316,14 +316,23 @@ void testLiteInterpreterModuleInfoBasic() {
   while (true) {
-    std::string module_info = bc.get_forward_method_debug_info(pc);
-    if (module_info == "<no module info for pc>") {
+    // std::string module_info = bc.get_forward_method_debug_info(pc);
+    // if (module_info == "<no module info for pc>") {
+    //   break;
+    // }
+    // if (!module_info.empty()) {
+    //   module_debug_info_set.insert(module_info);
+    // }
+    // ++pc;
```
Why leave these commented out?

kimishpatel commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -454,21 +488,125 @@ void testLiteInterpreterSequentialModuleInfo() {
+  std::unordered_set<std::string> module_debug_info_set;
+  size_t pc = 0;
+  while (true) {
+    // std::string module_info = bc.get_forward_method_debug_info(pc);
+    // if (module_info == "<no module info for pc>") {
+    //   break;
+    // }
+    // if (!module_info.empty()) {
+    //   module_debug_info_set.insert(module_info);
+    // }
+    // ++pc;
```
Remove commented code everywhere.

kimishpatel commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -454,21 +488,125 @@ void testLiteInterpreterSequentialModuleInfo() {
+    }
+  }
+
+  std::unordered_set<std::string> expected_result(
```
Thanks for adding this test. I think it might help to just add a comment explaining why 3 nodes. Each for each of the add. And each add occurs inside a different module.

kimishpatel commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -454,21 +488,125 @@ void testLiteInterpreterSequentialModuleInfo() {
+    }
+  }
+
+  std::unordered_set<std::string> expected_result(
```
Add a comment here as well to explain that this is the limitation of the current approach.

kimishpatel commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -77,9 +80,12 @@ void Function::set_register_size(size_t size) {
-  if (pc >= pc_to_module_debug_info_.size()) {
-    return "<no module info for pc>";
-  }
+  // if (pc >= pc_to_module_debug_info_.size()) {
+  //   return "<no module info for pc>";
+  // }
```
Remove commented code.

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

kimishpatel reviewed (changes requested):
Looks good mostly. 
Mainly two things:
- Remove commented code.
- Add comments in couple of tests to just help understand better.

And if anything left off of reconstruct_scopes, move to the base diff.

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

kimishpatel commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -65,10 +79,21 @@ void Function::set_register_size(size_t size) {
 }
 
+std::string Function::get_module_debug_info(size_t pc) {
+  // if (pc >= pc_to_module_debug_info_.size()) {
+  //   return "<no module info for pc>";
+  // }
```
Remove this.

kimishpatel reviewed (changes requested):
One more place where commented out code is left.

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

kimishpatel reviewed (approved):
Looks good. Let's make sure all the test pass. If the tests are failing, you can check master if they are failing on master here: https://ezyang.github.io/pytorch-ci-hud/build/pytorch-master. That way you know, you did not introduce the failure. For example pytorch_doc_test is failing.

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

kimishpatel commented:
Tai, @dreiss suggested that we always put the debug table and it can be empty. This ensures that, that particular position in the tuple always has debug info table. It will be empty when no debug info is saved. This way in future if another table is added, there is not ambiguity about what position in the table corresponds to what. If 5th element is debug table then it will always be that. It will be empty when no debug info present.
Note that this means going forward all models will be saved with debug table. 
But the model loading partly same. In that you still need to check # of elements in tuple to make sure we are backward compatible. Also if the debug table is empty you can ignore it instead of throwing an error.

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

iseeyuan commented:
> Tai, @dreiss suggested that we always put the debug table and it can be empty. This ensures that, that particular position in the tuple always has debug info table. It will be empty when no debug info is saved. This way in future if another table is added, there is not ambiguity about what position in the table corresponds to what. If 5th element is debug table then it will always be that. It will be empty when no debug info present.
> Note that this means going forward all models will be saved with debug table.
> But the model loading partly same. In that you still need to check # of elements in tuple to make sure we are backward compatible. Also if the debug table is empty you can ignore it instead of throwing an error.

As discussed offline, do we consider the option of putting it in a separate pickle file?

kimishpatel commented:
> > Tai, @dreiss suggested that we always put the debug table and it can be empty. This ensures that, that particular position in the tuple always has debug info table. It will be empty when no debug info is saved. This way in future if another table is added, there is not ambiguity about what position in the table corresponds to what. If 5th element is debug table then it will always be that. It will be empty when no debug info present.
> > Note that this means going forward all models will be saved with debug table.
> > But the model loading partly same. In that you still need to check # of elements in tuple to make sure we are backward compatible. Also if the debug table is empty you can ignore it instead of throwing an error.
> 
> As discussed offline, do we consider the option of putting it in a separate pickle file?

@iseeyuan, yes but I was thinking of doing it in another diff. What do you think? Of course this would be bc breaking if there are any models with debug info saved in the current way. But I think I will do a follow up diff soon after.

iseeyuan commented:
> > > Tai, @dreiss suggested that we always put the debug table and it can be empty. This ensures that, that particular position in the tuple always has debug info table. It will be empty when no debug info is saved. This way in future if another table is added, there is not ambiguity about what position in the table corresponds to what. If 5th element is debug table then it will always be that. It will be empty when no debug info present.
> > > Note that this means going forward all models will be saved with debug table.
> > > But the model loading partly same. In that you still need to check # of elements in tuple to make sure we are backward compatible. Also if the debug table is empty you can ignore it instead of throwing an error.
> > 
> > 
> > As discussed offline, do we consider the option of putting it in a separate pickle file?
> 
> @iseeyuan, yes but I was thinking of doing it in another diff. What do you think? Of course this would be bc breaking if there are any models with debug info saved in the current way. But I think I will do a follow up diff soon after.

My preference is to do it in this PR. The contents of the debug info can be kept. The major change would be another writearchive instead of directly embedding the table into bytecode.pkl.

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

kimishpatel commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -158,28 +160,31 @@ c10::IValue getFunctionTuple(
     }
+    auto module_debug_info = Table({{"module_debug_info", Tup(module_paths)}});
+    debug_info_elements.push_back(
+        Tup({func.qualname().qualifiedName(), module_debug_info}));
```
This is good that you are adding association information between bytecode.pkl and debug_info.pkl

kimishpatel commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -64,9 +64,11 @@ std::string getModulePath(Node* node) {
 }
 
-c10::IValue getFunctionTuple(
+void getFunctionTuple(
```
Just return a pair of IValue instead of passing a reference.

kimishpatel replied:
And actually for debug_info_elements return `c10::optional`. It kind of makes it clear that it is optional and you get only when you have save_debug_info_in_bytecode set. Also rename `save_debug_info_in_bytecode` to `save_mobile_debug_info` since we are not saving it in bytecode.pkl.

kimishpatel commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -188,13 +193,13 @@ void setstateTuple(
-      elements.push_back(
-          getFunctionTuple(module, setstate, save_debug_info_in_bytecode));
+      getFunctionTuple(
+          module, setstate, elements, debug_info_elements, save_debug_info_in_bytecode);
```
Let's function return pair. You can capture the return by reference but most likely they should be return value optimized even without capturing by ref. https://en.cppreference.com/w/cpp/language/copy_elision

kimishpatel commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -84,6 +84,7 @@ void print_unsupported_ops_and_throw(
 
 void parseMethods(
     const std::vector<IValue>& vals,
+    const std::vector<IValue>& debug_info_vals,
```
Maybe do optional here, so that instead of `bool has_debug_info = debug_info_vals.size() > 0;`, you can do `bool has_debug_info = debug_info_vals.has_value();`

kimishpatel reviewed (changes requested):
This is really good Tai. Thanks for making this change.

iseeyuan commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -898,19 +898,25 @@ void initJitScriptBindings(PyObject* module) {
           py::arg("filename"),
-          py::arg("_extra_files") = ExtraFilesMap())
+          py::arg("_extra_files") = ExtraFilesMap(),
+          py::arg("_save_debug_info_in_bytecode") = false)
```
Is this covered by tests from Python side? The tests could be added in test/mobile/test_lite_script_module.py.

taivu1998 replied:
I added Python tests. Since we have no access to the internal C++ object, I basically check if the outputs still match when we enable debug info saving and if module info strings are contained in the debug info buffer. Let me know if that suffices.

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

kimishpatel commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -26,9 +26,14 @@
 
 namespace {
+
+using func_tuple_type = std::pair<IValue, c10::optional<IValue>>;
```
I think it will be fine to not generate type alias, unless it is too long. This one is also not very informative.

kimishpatel commented on test/mobile/test_lite_script_module.py:
```diff
@@ -34,6 +34,58 @@ def forward(self, x):
+        torch.testing.assert_allclose(script_module_result, mobile_module_forward_result)
+
+        mobile_module_run_method_result = mobile_module.run_method("forward", input)
+        torch.testing.assert_allclose(script_module_result, mobile_module_run_method_result)
```
Thanks for adding the tests.

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -27,7 +27,8 @@
 //      ('types', ()),
-//      ('register_size', 2))))
+//      ('register_size', 2)
+//      ('module_debug_info', (top(A).foo(B).forward)))))
```
Should we also update this in the comment, that the debug info is in another .pkl file?

taivu1998 replied:
Thanks for catching this!

Commit:
Update on "Get, save, and load module information for each operator"


Differential Revision: [D22803740](https://our.internmc.facebook.com/intern/diff/D22803740)

[ghstack-poisoned]
